Added some scenarios to the architecture (#47)
ietf-wg-wimse/draft-ietf-wimse-arch: draft-ietf-wimse-arch.md

Reviewer: I've suggested harmonising diagram text and reduce "internal" vs "external" terminology when it comes to trust domains.

The certificate or WIT is presented during authentication, however the private key is kept secret and only used in cryptographic computation to prove that the presenter has access to the private key corresponding to the public key.

## Workload Identity System Scenarios

### Basic Workload Identity Scenario

~~~aasvg

>                        +------------------------------------------------------------+
>                        |                                                            |
>                        |                                                            |
>                     +-----+         +--------------+                                |
>                     |     |         |              |       (3)                      |
>                     |     |     +--->  Workload 1  +-------------------+            |
>                     |  G  |     |   |              |                   |            |
>                     |  a  |     |   +-------^------+                   |            |
>                     |  t  |     |           |                          |            |
> +------------+      |  e  |     |           | (1)                      |            |
> |            | (2)  |  w  | (3) |           |                          |            |
> | App Client +----->|  a  +-----+   +-------+-------+                  |            |
> |            |      |  y  |         |               |          +-------+------+     |
> +------------+      |     +-----+   | CA/Credential |    (1)   |              |     |
>                     |  S  | (4) |   |    Service    +---------->   Workload 3 |     |
>                     |  e  |     |   |               |          |              |     |
>                     |  r  |     |   +-------+-------+          +-------+------+     |
>                     |  v  |     |           |                          |            |
>                     |  i  |     |           | (1)                      |            |
>                     |  c  |     |           |                          |            |
>                     |  e  |     |   +-------v------+                   |            |
>                     |     |     |   |              |      (4)          |            |
>                     |     |     +--->  Workload 2  +-------------------+            |
>                     |     |         |              |                                |
>                     +-----+         +--------------+                                |
>                        |                                                            |
>                        |                                                            |
>                        |                                  Internal Trust Domain     |

Reviewer: ```suggestion
                       +-------------------------------Trust Boundary---------------+
                       |                                                            |
                       |                                                            |
                    +-----+         +--------------+                                |
                    |     |         |              |       (3)                      |
                    |     |     +--->  Workload 1  +-------------------+            |
                    |  G  |     |   |              |                   |            |
                    |  a  |     |   +-------^------+                   |            |
                    |  t  |     |           |                          |            |
+------------+      |  e  |     |           | (1)                      |            |
|            | (2)  |  w  | (3) |           |                          |            |
| App Client +----->|  a  +-----+   +-------+-------+                  |            |
|            |      |  y  |         |               |          +-------+------+     |
+------------+      |     +-----+   | CA/Credential |    (1)   |              |     |
                    |  S  | (4) |   |    Service    +---------->   Workload 3 |     |
                    |  e  |     |   |               |          |              |     |
                    |  r  |     |   +-------+-------+          +-------+------+     |
                    |  v  |     |           |                          |            |
                    |  i  |     |           | (1)                      |            |
                    |  c  |     |           |                          |            |
                    |  e  |     |   +-------v------+                   |            |
                    |     |     |   |              |      (4)          |            |
                    |     |     +--->  Workload 2  +-------------------+            |
                    |     |         |              |                                |
                    +-----+         +--------------+                                |
                       |                                                            |
                       |                                                            |
                       |                                                            |
```

Harmonise diagrams

                       +------------------------------------------------------------+
~~~~
{: #arch-basic title="Basic example workload application system."}

> The above diagram presents a basic workload application system.  The large box represents a trust boundary within which the workload application is hosted.  Within this example

Reviewer: ```suggestion
The above diagram presents a basic workload application system.  The large box represents a trust domain within which the workload application is hosted.  Within this example
```

Let's keep terms consistent

there are three workloads, a gateway, that accepts external clients and a CA/credential service that issues workload identity credentials for the trust domain.  External
to the workload application system there is an application client that calls APIs on workloads.

Here is a brief summary of each component

* Trust Domain

> The large box represents the internal trust domain of the application that is composed of several workloads.  A trust domain may have a more complex internal structure with more workloads, multiple gateways, internal infrastructure, and other services.

Reviewer: ```suggestion
The large box represents a trust domain of the application that is composed of several workloads.  A trust domain may have a more complex internal structure with more workloads, multiple gateways, internal infrastructure, and other services.
```

"Internal" means different things for different people

* Workload

Three workloads are shown.  Each workload is an instance of running software executing for a specific purpose.  Workloads obtain their identity credentials from a Credentials Service (1) and use them to authenticate to other workloads and systems in the process
of sending and receiving requests to and from external systems or other internal workloads.

* Gateway Service

A gateway service typically acts as an intermediary between the internal application trust domain and external systems. The gateway is responsible for ensuring appropriate isolation between external and internal domains. It also routes incoming requests to the correct workload.
The gateway may also handle authentication, token exchange, and token transformation.

* CA/Credential Service

> In this diagram the token/Credential service is a service responsible for issuing workload identities to workloads in the internal trust domain. The credentials are often X.509 based or JWT based.

Reviewer: ```suggestion
In this diagram the token/Credential service is a service responsible for issuing workload identities to workloads in the same trust domain. The credentials are often X.509 based or JWT based.
```
Same as above

* (4) API request to workload 2

Similarly to the previous flow, the gateway may determine that for another API call, the application client's request needs to be handled by workload 2. The case behaves the same as the previous flow except that the gateway may need to authenticate workload 2 before forwarding traffic to it. Workload 3 will then authorize and audit the request based on the authenticated identity of workload 2. Workload 2 and workload 1 may be authorized to use different APIs on workload 3. If workload 1 or 2 makes an API request that it is not authorized for, then workload 3 will reject the request.

### Context and workload Identity

~~~aasvg

>                  +----------------------------------------------+
>                  |                                              |
>                  |                      +----------------+      |
>               +--+--+                   |                |      |
>               |     +------------------>|    Context     |      |
>               |     |         (3)       |                |      |
>               |     |<------------------+    Service     |      |
>               |     |         (c)       |                |      |
>               |  G  |                   +----------------+      |
>               |  a  |                                           |
>    (1)        |  t  |                                           |
> +-------+     |  e  |      +------------+      +-------------+  |
> |   App | (2) |  w  | (4)  |            |  (5) |             |  |
> | Client+---->|  a  +----->| workload 1 +----->|  workload 2 |  |
> +-------+ (a) |  y  | (c)  |            |  (c) |             |  |
>               |     |      +------------+      +-------------+  |
>               |  S  |                                           |
>               |  e  |                                           |
>               |  r  |                                           |
>               |  v  |                                           |
>               |  i  |                                           |
>               |  c  |                                           |
>               |  e  |                                           |
>               |     |                                           |
>               +--+--+                                           |
>                  |                                              |
>                  |                                              |
>                  |                       Internal Trust Domain  |

Reviewer: ```suggestion
                 +-----------------------Trust Boundary---------+
                 |                                              |
                 |                      +----------------+      |
              +--+--+                   |                |      |
              |     +------------------>|    Context     |      |
              |     |         (3)       |                |      |
              |     |<------------------+    Service     |      |
              |     |         (c)       |                |      |
              |  G  |                   +----------------+      |
              |  a  |                                           |
   (1)        |  t  |                                           |
+-------+     |  e  |      +------------+      +-------------+  |
|   App | (2) |  w  | (4)  |            |  (5) |             |  |
| Client+---->|  a  +----->| workload 1 +----->|  workload 2 |  |
+-------+ (a) |  y  | (c)  |            |  (c) |             |  |
              |     |      +------------+      +-------------+  |
              |  S  |                                           |
              |  e  |                                           |
              |  r  |                                           |
              |  v  |                                           |
              |  i  |                                           |
              |  c  |                                           |
              |  e  |                                           |
              |     |                                           |
              +--+--+                                           |
                 |                                              |
                 |                                              |
                 |                                              |
```

Same as above

The gateway may use alternative mechanisms to get the internal security context information (c).

* (4) Forwarding Request to Workload
The gateway forwards the request along with the context information (c) to the appropriate workload. A bearer token, such as an access token (a), is not usually forwarded as it is only meant for external access. The workload uses information in the context token in applying authorization policy to the application client's request.
If the workload does not receive a context token, then it will deny requests that rely on information from the token.

* (5) Making Additional Workload Originated Requests
The workload may need to make requests of other workloads. When making these requests, the workload includes the context information so Workload 2 can authorize and audit the request. Workload 2 may have a policy requiring Workload 1 to authenticate its service identity and provide valid context information (c) to access certain APIs.

### Cross-Domain Communication
~~~aasvg

>                  +----------------------------------------------+
>                  |                                              |
>               +--+--+                                           |
>               |     |                                           |
>               |  G  |                                           |
>               |  a  |                                           |
>               |  t  |                                           |
> +-------+     |  e  |      +------------+      +-------------+  |
> |   App | (1) |  w  | (2)  |            |  (4) |             |  |
> | Client+---->|  a  +----->| workload 1 +----->|  workload 2 |  |
> +-------+ (a) |  y  | (c)  |            |  (c) |             |  |
>               |     |      +-+---------++      ++--------+---+  |
>               |  S  |        |                  |        |      |
>               |  e  |     (3)|               (5)|(c)     |      |
>               |  r  |        |                  |        |      |
>               |  v  |        |       +-+--------++       |      |
>               |  i  |        |       |   Token   |       |      |
>               |  c  |        |       |           |    (6)|(t)   |
>               |  e  |        |       |  Service  |       |      |
>               |     |        |       +-----------+       |      |
>               +--+--+        |                           |      |
>                  |           |                           |      |
>                  |           |                           |      |
>                  |           |             Internal Trust|Domain|

Reviewer: ```suggestion
                 +--------------------------Trust Boundary------+
                 |                                              |
              +--+--+                                           |
              |     |                                           |
              |  G  |                                           |
              |  a  |                                           |
              |  t  |                                           |
+-------+     |  e  |      +------------+      +-------------+  |
|   App | (1) |  w  | (2)  |            |  (4) |             |  |
| Client+---->|  a  +----->| workload 1 +----->|  workload 2 |  |
+-------+ (a) |  y  | (c)  |            |  (c) |             |  |
              |     |      +-+---------++      ++--------+---+  |
              |  S  |        |                  |        |      |
              |  e  |     (3)|               (5)|(c)     |      |
              |  r  |        |                  |        |      |
              |  v  |        |       +-+--------++       |      |
              |  i  |        |       |   Token   |       |      |
              |  c  |        |       |           |    (6)|(t)   |
              |  e  |        |       |  Service  |       |      |
              |     |        |       +-----------+       |      |
              +--+--+        |                           |      |
                 |           |                           |      |
                 |           |                           |      |
                 |           |                           |      |
```

Same as above

                 +-----------+---------------------------+------+
                             |                           |
                             |                           |
                             |                           |
                             |                           |
                             |                           |
                     +-------v--------+     +------------v--+
                     | Infrastructure |     |    External   |
                     |                |     |               |
                     |    Service     |     |    Service    |
                     +----------------+     +---------------+

~~~~
{: #arch-cross title="External request workload application system."}

In many applications workloads must make requests of infrastructure that operates as a different trust domain or external services that are in a separate trust domain. Figure {{arch-cross}} shows this scenario. The scenario shows some new components described below.
Components and interactions from previous scenarios are still relevant to this example, but are omitted for simplicity.

> * Token Service - the token service is responsible for exchanging information that is internal to the system such as service identity and/or security context information for a token that can be presented to an external service to gain access to infrastructure or an external service.  Note that in some cases the token service may reside in the external trust domain or there may be token services in both the internal trust domain and the external trust domain.  In some cases the workload will need to contact both the internal token service and an external token service in order to gain access to infrastructure or external service.

Reviewer: ```suggestion
* Token Service - the token service is responsible for exchanging information that is internal to the system such as service identity and/or security context information for a token that can be presented to an external service to gain access to infrastructure or an external service.  Note that in some cases the token service may reside in a different trust domain or there may be token services in multiple trust domains.  In some cases the workload will need to contact token services in various domains in order to gain access to infrastructure or external service.
```

Same as above
